_posts/2021-01-13-hapi-fhir.md

```diff
@@ -0,0 +1,454 @@
+
+  
+
+```Java
+import ca.uhn.fhir.context.FhirContext;
+import ca.uhn.fhir.rest.annotation.*;
+import ca.uhn.fhir.rest.api.MethodOutcome;
+import ca.uhn.fhir.rest.param.DateRangeParam;
+import ca.uhn.fhir.rest.param.StringParam;
+import ca.uhn.fhir.rest.param.TokenParam;
+import ca.uhn.fhir.rest.server.IResourceProvider;
+import com.example.fhirexample.dao.patient.IPatient;
+import org.hl7.fhir.instance.model.api.IBaseResource;
+import org.hl7.fhir.r4.model.IdType;
+import org.hl7.fhir.r4.model.OperationOutcome;
+import org.hl7.fhir.r4.model.Patient;
+import org.hl7.fhir.r4.model.Resource;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+import javax.servlet.http.HttpServletRequest;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@Component
+public class PatientProvidor implements IResourceProvider {
+
+    @Autowired
+    private FhirContext ctx;
+
+    @Autowired
+    private IPatient patientDao;
+
+    private Map<String, Patient> patientMap = new HashMap<String, Patient>();
+
+    public  PatientProvidor() {
+    
+    }   
+    
+    @Override
+    public Class<? extends IBaseResource> getResourceType() {
+        return  Patient.class;
+    }
+
+    @Search
+    public List<Resource> searchPatient(HttpServletRequest request,
+                                        @OptionalParam(name= Patient.SP_BIRTHDATE) DateRangeParam birthDate,
+                                        @OptionalParam(name = Patient.SP_FAMILY) StringParam familyName,
+                                        @OptionalParam(name= Patient.SP_GENDER) StringParam gender,   
+                                        @OptionalParam(name= Patient.SP_GIVEN) StringParam givenName,
+                                        @OptionalParam(name = Patient.SP_IDENTIFIER) TokenParam identifier,
+                                        @OptionalParam(name= Patient.SP_NAME) StringParam name,
+                                        @OptionalParam(name = Patient.SP_RES_ID) TokenParam resid) {
+
+        List<Resource> results = patientDao.search(ctx, birthDate, familyName, gender, givenName, identifier, name);
+        return results;
+    }
+
+    @Read()
+    public Patient read(@IdParam IdType theId) {
+
+        Patient patient = patientDao.read(ctx,theId);
+        return patient;
+    }
+
+    @Create()
+    public MethodOutcome createPatient(HttpServletRequest theRequest, @ResourceParam Patient patient) {
+
+        MethodOutcome  method = new  MethodOutcome();
+        method.setCreated(true);
+        OperationOutcome  opOutcome = new  OperationOutcome();
+        method.setOperationOutcome(opOutcome)
+       
+        Patient  mongoPatient = patientDao.create(ctx, patient);
+        method.setId(mongoPatient.getIdElement());
+        method.setResource(mongoPatient);
```

Comment:
Hmmm, not sure about this ... are you sure that you need the dao class to a setRESOUCE() method? I would expect to pass the FHIR Patient resource class here. Can you please double check?

Reply:
Bad variable naming.  You get a FhirPatient object back from the Patient DAO create.
I will change the variable

Reply:
just remembered that I still need to change this